### ISSUE TITLE
rule no-undef false positive when using nested destructuring arguments  

### ISSUE THREAD

**Issue by mwain** [eslint/eslint#6003]
**What version of ESLint are you using?**
2.8.0

**What parser (default, Babel-ESLint, etc.) are you using?**
default

**What did you do? Please include the actual source code causing the issue.**
Using Node v6 - es6 nested destructuring arguments undef is erroring.

```
/* eslint-env es6 */
/* eslint no-undef: "error" */

function myFunc({foo, bar}) { return bar }
function myFunc({foo, bar = {a, b}}) { return bar }
```
**What did you expect to happen?**
Should be valid

**What actually happened? Please include the actual, raw output from ESLint.**
`a` and `b` is throwing an error

```
  5:30  error  'a' is not defined  no-undef
  5:33  error  'b' is not defined  no-undef
```

**Comment by mysticatea** [comment]
Thank you for this issue.

This is correct behavior.
The `{a, b}` is an object literal: `{a: a, b: b}`. So it needs variables `a` and `b`.

**Comment by mwain** [comment]
`function myFunc({foo, bar = {a: a, b: b}}) { return bar }`
Still causes the error. 

`function myFunc({foo, bar = {a, b}}) { return bar }`
Is valid and works.

**Comment by mysticatea** [comment]
Yes, it's valid syntax in sloppy mode.
But `a` and `b` are references to undefined variables.

**Comment by mwain** [comment]
Its object destructuring parameters, with no defaults.
`a` and `b` wont be defined as this is in the function definition.

Am i missing something?

**Comment by mysticatea** [comment]
Do you want the following code?

```js
function myFunc({foo, bar: {a, b}}) { return bar }
```

**Comment by mwain** [comment]
But now `bar` is not defined? 
I would prefer `bar` to be an object still with `a` and `b` properties of it.

**Comment by mysticatea** [comment]
Yes, `bar` is not defined in this case by ECMAScript spec.

**Comment by mysticatea** [comment]
You can:

```js
function myFunc({foo, bar}) {
    let {a, b} = bar;
    return bar;
}

**Comment by mwain** [comment]
Hmm! Interesting i was following an example from https://developer.mozilla.org/en/docs/Web/JavaScript/Reference/Operators/Destructuring_assignment section 'Setting a function parameter's default value'

And thought it would work without the defaults being set, as i was testing it
`function myFunc({foo, bar = {a, b}}) { return bar }`
It worked as i expected, `bar` was an object with `a` and `b` properties of it.

**Comment by mysticatea** [comment]
```js
function drawES6Chart({size = 'big', cords = { x: 0, y: 0 }, radius = 25} = {}) {
  console.log(size, cords, radius);
  // do some chart drawing
}
```

`{ x: 0, y: 0 }` is an object literal, it's the default value of `cords`. `cords.x` is `0`, `cords.y` is `0`.
If the object literal is `{ x: x, y: y }`, it can be written as `{x, y}` also (it's shorthand properties).
In this case, `x` and `y` are global variables.

**Comment by michaelficarra** [comment]
@mwain what you're looking for is known as as-patterns, and ECMAScript destructing does not have them.
